packages/taiko-client/prover/proof_submitter/proof_buffer.go

```diff
@@ -0,0 +1,92 @@
+package submitter
+
+import (
+	"errors"
+	"sync"
+
+	producer "github.com/taikoxyz/taiko-mono/packages/taiko-client/prover/proof_producer"
+)
+
+var (
+	errBufferOverflow = errors.New("proof buffer overflow")
+	errNotEnoughProof = errors.New("not enough proof")
+)
+
+type ProofBuffer struct {
+	MaxLength uint64
+	buffer    []*producer.ProofWithHeader
+	mutex     sync.Mutex
+}
+
+func NewProofBuffer(maxLength uint64) *ProofBuffer {
+	return &ProofBuffer{
+		buffer:    make([]*producer.ProofWithHeader, 0, maxLength),
+		MaxLength: maxLength,
+	}
+}
+
+func (pb *ProofBuffer) Write(item *producer.ProofWithHeader) (int, error) {
+	pb.mutex.Lock()
+	defer pb.mutex.Unlock()
+
+	if len(pb.buffer)+1 > int(pb.MaxLength) {
+		return len(pb.buffer), errBufferOverflow
+	}
+
+	pb.buffer = append(pb.buffer, item)
+	return len(pb.buffer), nil
+}
+
+func (pb *ProofBuffer) Read(length int) ([]*producer.ProofWithHeader, error) {
+	pb.mutex.Lock()
```

Comment:
Let's just use read lock here？

Reply:
It seems that there is no need to add a lock in the read function?